### ISSUE TITLE
Copying courses may result in badly named groups

### ISSUE THREAD

**Issue by darius-m** [microsoft/o365-moodle#1448]
It appears that when Moodle tries to copy a course, using the copy course feature, it will first [create a course](https://github.com/moodle/moodle/blob/master/backup/util/ui/classes/copy/copy.php#L147) named `Course copying in progress` that will be later be filled with the data from the base course / the UI form.

Because the course name is not correct from the beginning, having the o365 plugin set to sync all courses automatically, means that the new groups / teams will use the default `Course copying in progress` name, and worse yet, they will not update when the copy process finishes (I do not believe any events are generated by this process). Considering that the skeleton course [has its visibility set to 0](https://github.com/moodle/moodle/blob/master/backup/util/dbops/restore_dbops.class.php#L1826) until the copy is completed, a possible fix would be to disable the o365 sync for non-visible courses (with the obvious downside of the synchronization not being performed until the course is made visible). 

Unless there is another point in the code where the sync could be performed, that the fix would be to change the [query for non-synced courses](https://github.com/microsoft/o365-moodle/blob/master/local/o365/classes/feature/usergroups/coursegroups.php#L110) to look like this:
```php
$sql = 'SELECT crs.*
          FROM {course} crs
     LEFT JOIN {local_o365_objects} obj ON obj.type = ? AND obj.subtype = ? AND obj.moodleid = crs.id
          WHERE obj.id IS NULL AND crs.id != ? and crs.visible != 0';
```

**Comment by weilai-irl** [comment]
Hi @darius-m,

Thank you for your suggestion. This is a very valid point, and your suggestion makes sense.

I do have a concern, though - the visibility of a course is not 100% depending on the "visible" field in the database, other settings, such as visibility of the course category, may have impact on it as well, so the solution may not be as easy as a simple query update. Also, we may need to consider adding observer functions to create group/team for courses when their visibility gets updated. The observer function will be likely on course updated events and course category updated events.

This will be part of the release in January or March 2021.

Regards,
Lai

**Comment by darius-m** [comment]
Hello, @weilai-irl 

The suggestion regarding visibility wasn't really meant to be aimed at having courses which are set to not visible, not create groups. It is more of a workaround in case the course is copied or restored, since the process seems to first create a course which has its visibility set to 0 with a default name, update the course information and in the end makes it visible.

I do understand, however, that since this is a common functionality for courses, both restored and not, the behavior may become confusing for the users.

Thank you for your input!
Darius

**Comment by weilai-irl** [comment]
Hi @darius-m,

After some rethink, I agree that the approach you suggested would be the easiest and safest to prevent courses in the progress of restore or copy to be synced to Teams or groups. I have clearly been overthinking in my last reply.

I have implemented exactly as you suggested, and the changes will be included in the next release.

Regards,
Lai

**Comment by weilai-irl** [comment]
Hi @darius-m,

The suggested change has been included in the latest release (3.9.4 and 3.10.1) of the plugins. Please check them out.

Thank you for your contribution.

Regards,
Lai
